File: n-queens.c
void queens(index i)
{
	index j;
	if (promising(i))
		if (i == n) #n개의 퀸을 다 놓았다는 뜻
			print col[1] through col[n];
		else #n개의 퀸 다 안놓은 상황이라 
			for (j=1; j<=ㅜ; j++) {
			col[i + 1] = j
			queens(i + 1);
			}
}

#i가 promising한가에 대한 검사
bool promising(index i)
{
	index k;
	bool switch;
	k = 1;
	switch = ture;
	while (k < i && switch) {
		if (col[i]==col[k] || abs(col[i]-col[k]) == i-k) #컬럼값이 동일 || 대각선의경우는 nonpromising!
			switch = false;
		k++;
	}
	return switch;
}

#호출 
queens(0)

File: n-queens.py
""" n-queens문제의 해결
기본 가정 : 같은 행에는 퀸을 놓지 않고
구현시 : 같은 열이나, 대각선에 놓이는지를 확인해 본다"""

#column체크하기
#column[i]==column[k]

#diagonal(대각선)체크하기
#절대값이 대각선임
#abs(col[i] - col[k]) == i - k

def promising(i, col):
    k = 1
    flag = True
    while (k < i and flag):
        if (col[i] == col[k]) or abs(col[i]-col[k]) == (i-k):
            flag = False
        k += 1
    return flag


def n_queens(i, col):
    n = len(col) - 1
    if (promising(i, col)):
        if i == n:
            print(col[1:n+1])
        else:
            for j in range(1, n+1):
                col[i+1] = j
                n_queens(i+1, col)

n = 4
col = [0] * (n+1)
n_queens(0, col)
